### PR TITLE
#244: Rework cameras to avoid releasing it when checking connection

### PR DESCRIPTION
**Pull request by anxuae**
The objective of this PR is to keep camera proxy opened when discovered, it permits to keep it busy and avoid detection by another drivers library : fix for #244

### CODE REVIEW

**Review by werdeil — COMMENTED** [review]
Small naming comments only as expected

File: pibooth/camera/__init__.py [hunk 1]
```diff
@@ -1,40 +1,52 @@
...
 
-    The priority order is chosen in order to have best rendering during preview
-    and to take captures.
+def find_camera():
+    """Initialize the camera depending of the connected one. The priority order
+    is chosen in order to have best rendering during preview and to take captures.
+    The gPhoto2 camera is first (drivers most restrictive) to avoid connection
+    concurence in case of DSLR compatible with OpenCV.
     """
-    if rotation not in (0, 90, 180, 270):
```
[the remark this answers is not in the document]
Comment (anxuae) [reply pibooth/camera/__init__.py:17]: Yes re-add it in the BaseCamera.initialize()

File: pibooth/camera/__init__.py [hunk 2]
```diff
@@ -1,40 +1,52 @@
...
+from pibooth.camera.rpi import RpiCamera, find_rpi_camera
+from pibooth.camera.gphoto import GpCamera, find_gp_camera
+from pibooth.camera.opencv import CvCamera, find_cv_camera
 from pibooth.camera.hybrid import HybridRpiCamera, HybridCvCamera
 
 
-def get_camera(iso, resolution, rotation, flip, delete_internal_memory):
-    """Initialize the camera depending of the connected one. If a gPhoto2 camera
-    is used, try to kill any process using gPhoto2 as it may block camera access.
+def _clean(rpi_cam, gp_cam, cv_cam):
```
[the remark this answers is not in the document]
Comment (anxuae) [reply pibooth/camera/__init__.py:10]: renamed "proxy" suffixe

File: pibooth/camera/__init__.py [hunk 3]
```diff
@@ -1,40 +1,52 @@
...
+    is chosen in order to have best rendering during preview and to take captures.
+    The gPhoto2 camera is first (drivers most restrictive) to avoid connection
+    concurence in case of DSLR compatible with OpenCV.
     """
-    if rotation not in (0, 90, 180, 270):
-        raise ValueError("Invalid camera rotation value '{}' (should be 0, 90, 180 or 270)".format(rotation))
-    if gp_camera_connected() and rpi_camera_connected():
+    rpi_cam = find_rpi_camera()
+    gp_cam = find_gp_camera()
+    cv_cam = find_cv_camera()
```
[the remark this answers is not in the document]
Comment (anxuae) [reply pibooth/camera/__init__.py:29]: renamed "proxy" suffixe

File: pibooth/camera/base.py [hunk 4]
```diff
@@ -9,17 +9,35 @@
...
-        self.delete_internal_memory = delete_internal_memory
+        self.capture_flip = flip
         if not isinstance(iso, (tuple, list)):
             iso = (iso, iso)
-        self.iso_preview, self.iso_capture = iso
+        self.preview_iso, self.capture_iso = iso
+        self.delete_internal_memory = delete_internal_memory
+        self._initialize()
+
+    def _initialize(self):
```
[the remark this answers is not in the document]
Comment (anxuae) [reply pibooth/camera/base.py:37]: ok

File: pibooth/camera/opencv.py [hunk 5]
```diff
@@ -9,24 +9,33 @@
...
-from pibooth.utils import PoolingTimer, memorize, LOGGER
+from pibooth.utils import PoolingTimer, LOGGER
 from pibooth.language import get_translated_text
 from pibooth.camera.base import BaseCamera
 
 
-@memorize
-def cv_camera_connected():
-    """Return True if a camera compatible with OpenCV is found.
+def find_cv_camera(port=None):
```
[the remark this answers is not in the document]
Comment (anxuae) [reply pibooth/camera/opencv.py:17]: changed

File: pibooth/camera/rpi.py [hunk 6]
```diff
@@ -8,26 +8,29 @@
...
     picamera = None  # picamera is optional
-from pibooth.utils import memorize
 from pibooth.language import get_translated_text
 from pibooth.camera.base import BaseCamera
 
 
-@memorize
-def rpi_camera_connected():
-    """Return True if a RPi camera is found.
+def find_rpi_camera(port=None):
```
[the remark this answers is not in the document]
Comment (anxuae) [reply pibooth/camera/rpi.py:15]: changed

**Pull request merged**